// easy-currencies/dist/parts/chainer.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var converter_1 = require("../converter");
var await_to_js_1 = __importDefault(require("await-to-js"));
/**
 * Chained converter.
 * It defaults to the basic API provider, and does not require initialization.
 *
 * @example
 *  let value = await Convert(15).from("USD").to("EUR");
 *
 * @export
 * @param {number} amount - amount of currency to convert
 * @returns
 */
function Chainer(amount) {
    if (amount === void 0) { amount = undefined; }
    var _currentAmount = amount;
    var _currentFrom = undefined;
    var _currentTo = undefined;
    var _currentRates = undefined;
    // local converter
    var c = new converter_1.Converter();
    /**
     *  Return object construction, prepared for chaining.
     */
    var ob = {
        from: _from,
        to: _to,
        fetch: _fetch,
        get rates() {
            return _currentRates;
        },
        amount: _amount
    };
    /**
     * Chain member that fetches and caches the rates for the given currency.
     *
     * @returns {Promise<any>}
     */
    function _fetch() {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, c.getRates(_currentFrom, "", true)];
                    case 1:
                        // fetching rates for the base currency
                        _currentRates = _a.sent();
                        return [2 /*return*/, ob];
                }
            });
        });
    }
    // returning chainable
    return ob;
    /**
     * Chain member that sets the base currency
     *
     * @param {string} from
     * @returns chainable object
     */
    function _amount(val) {
        _currentAmount = val;
        return ob;
    }
    /**
     * Chain member that sets
     *
     * @param {string} from
     * @returns chainable object
     */
    function _from(from) {
        _currentFrom = from;
        return ob;
    }
    /**
     * Final chain member (ends chain)
     *
     * @param {string} to - final currency
     * @returns {number} - final converted amount
     */
    function _to(to) {
        return __awaiter(this, void 0, void 0, function () {
            var _a, err, r;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _currentTo = to;
                        return [4 /*yield*/, await_to_js_1.default(c.convert(_currentAmount, _currentFrom, _currentTo, _currentRates))];
                    case 1:
                        _a = _b.sent(), err = _a[0], r = _a[1];
                        if (err) {
                            throw err;
                        }
                        return [2 /*return*/, r];
                }
            });
        });
    }
}
exports.Chainer = Chainer;
