// easy-currencies/dist/parts/config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var providers_1 = require("./providers");
var utils_1 = require("./utils");
/**
 * Config object that initializes with configuration data
 * passed in by the user.
 *
 * @export
 * @class Config
 */
var Config = /** @class */ (function () {
    /**
     * Creates an instance of Config.
     * @param {(object | undefined)} config
     * @memberof Config
     */
    function Config() {
        var _this = this;
        var config = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            config[_i] = arguments[_i];
        }
        /**
         * Adds a single new, user-defined provider to the list of providers.
         *
         * @param {string} name - the new provider name
         * @param {Provider} provider - the new provider object
         * @param {boolean} [setActive=false] - should the new provider(s) be prioritized
         * @memberof Config
         */
        this.add = function (name, provider, setActive) {
            if (setActive === void 0) { setActive = false; }
            _this.addMultiple([{ name: name, provider: provider }], setActive);
        };
        /**
         * Adds multiple new, user-defined provider to the list of providers.
         *
         * @param {UserDefinedProvider[]} providers - providers to be added
         * @param {boolean} [setActive=false] - should the new provider(s) be prioritized
         * @memberof Config
         */
        this.addMultiple = function (newProviders, setActive) {
            if (setActive === void 0) { setActive = false; }
            // Duplicate check
            newProviders.forEach(function (p) {
                if (!utils_1.checkIfUserDefinedProvider(p)) {
                    throw "Invalid provider format!";
                }
                if (providers_1.providers[p.name]) {
                    throw "A provider by this name is already registered!";
                }
                providers_1.providers[p.name] = p.provider;
            });
            // Adding provider to active providers
            _this.addProviders(newProviders.map(function (p) { return p.provider; }), setActive);
        };
        /**
         * Removes a specific provider
         * @param {Provider} provider - provider to be removed
         * @memberof Config
         */
        this.remove = function (provider) {
            _this._active = _this._active.filter(function (p) { return p != provider; });
        };
        this._active = resolveProviders.apply(void 0, config);
        // adding default fallback
        this.addProviders([providers_1.providers["ExchangeRatesAPI"]], false);
    }
    Object.defineProperty(Config.prototype, "providers", {
        /**
         * Provider getter
         *
         * @returns {Provider[]}
         * @memberof Config
         */
        get: function () {
            return this._active;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Provider setter (adder)
     *
     * @param {Provider[]} providers - providers to be added
     * @param {boolean} [setActive=false] - should the new provider(s) be prioritized
     * @returns {void}
     * @memberof Config
     */
    Config.prototype.addProviders = function (providers, setActive) {
        var _a, _b;
        var _this = this;
        providers = providers.filter(function (p) {
            return !_this._active.find(function (a) { return a == p; });
        });
        if (setActive) {
            (_a = this._active).unshift.apply(_a, providers);
            return;
        }
        (_b = this._active).push.apply(_b, providers);
        return;
    };
    /**
     * Returns the current provider
     *
     * @returns {Provider} - current provider
     * @memberof Config
     */
    Config.prototype.activeProvider = function () {
        return this._active[0];
    };
    return Config;
}());
exports.Config = Config;
/**
 * Config resolver that normalizes configuration input into the config interface
 *
 * @export
 * @param {(object | undefined)} configuration - initial configuration
 * @returns {config} - normalized configuration object
 */
function resolveProviders() {
    var configuration = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        configuration[_i] = arguments[_i];
    }
    // resolve default if none provided.
    if (typeof configuration === "undefined" || !configuration.length) {
        return [providers_1.providers.ExchangeRatesAPI];
    }
    // checking for incorrect config types
    if (typeof configuration[0] !== "object" &&
        typeof configuration[0] !== "undefined" &&
        typeof configuration[0] !== "string") {
        throw "You must either supply nothing or a config object (see the 'config' section to see the different APIs that can be used)";
    }
    // returning single provider
    if (typeof configuration[0] === "string") {
        // constructing in initializationConfig object from string values
        return [providers_1.resolveProvider({ name: configuration[0], key: configuration[1] })];
    }
    // configuration is an array of providers
    // casting
    var c = configuration;
    // resolving all providers
    return c.map(function (provider) { return providers_1.resolveProvider(provider); });
}
exports.resolveProviders = resolveProviders;
