// easy-currencies/dist/parts/providers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * A function that constructs provider based on raw input data.
 *
 * @export
 * @param {*} provider object containing provider name and api key
 * @returns {Provider} constructed provider
 */
function resolveProvider(provider) {
    var p = exports.providers[provider.name];
    if (!p) {
        throw "No provider with this name. Please use a provider from the supported providers list.";
    }
    // attaching key
    p.key = provider.key;
    return p;
}
exports.resolveProvider = resolveProvider;
/**
 * Provider map initialization
 */
exports.providers = {
    ExchangeRatesAPI: {
        endpoint: {
            base: "https://api.exchangeratesapi.io/latest",
            single: "?base=%FROM%&symbols=%TO%",
            multiple: "?base=%FROM%"
        },
        key: null,
        handler: function (data) {
            return data.rates;
        },
        errors: { 400: "Malformed query." },
        errorHandler: function (data) {
            return data.status;
        }
    },
    CurrencyLayer: {
        endpoint: {
            base: "http://apilayer.net/api/live?access_key=%KEY%",
            single: "&source=%FROM%",
            multiple: "&source=%FROM%&currencies=%TO%"
        },
        key: undefined,
        handler: function (data) {
            var map = {};
            Object.keys(data.quotes).map(function (key) {
                map[key.slice(3)] = data.quotes[key];
            });
            return map;
        },
        errors: {
            105: "A paid plan is required in order to use CurrencyLayer (base currency use not allowed)",
            101: "Invalid API key!",
            201: "Invalid base currency."
        },
        errorHandler: function (data) {
            return data.error ? data.error.code : null;
        }
    },
    OpenExchangeRates: {
        endpoint: {
            base: "https://openexchangerates.org/api/latest.json?app_id=%KEY%",
            single: "&base=%FROM%",
            multiple: "&base=%FROM%"
        },
        key: undefined,
        handler: function (data) {
            return data.rates;
        },
        errors: {
            401: "Invalid API key!"
        },
        errorHandler: function (data) {
            return data.status;
        }
    },
    AlphaVantage: {
        endpoint: {
            base: "https://www.alphavantage.co/query?function=CURRENCY_EXCHANGE_RATE&apikey=%KEY%",
            single: "&from_currency=%FROM%&to_currency=%TO%",
            multiple: ""
        },
        key: undefined,
        handler: function (data) {
            var map = {};
            var o = data[Object.keys(data)[0]];
            map[o["3. To_Currency Code"]] = o["5. Exchange Rate"];
            return map;
        },
        errors: {
            503: "Invalid API key or Malformed query."
        },
        errorHandler: function (data) {
            return data["Error Message"] ? 503 : false;
        }
    },
    Fixer: {
        endpoint: {
            base: "http://data.fixer.io/api/latest?access_key=%KEY%",
            single: "&base=%FROM%&symbols=%TO%",
            multiple: "&base=%FROM%"
        },
        key: undefined,
        handler: function (data) {
            return data.rates;
        },
        errors: {
            105: "A paid plan is required in order to use Fixer.io (base currency use not allowed)",
            101: "Invalid API key!",
            201: "Invalid base currency."
        },
        errorHandler: function (data) {
            return data.error ? data.error.code : null;
        }
    }
};
